janus_server/src/aggregator.rs

```diff
@@ -933,6 +1010,171 @@ impl VdafOps {
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+        }
+    }
+
+    async fn handle_aggregate_share_generic<A, E>(
+        datastore: &Datastore,
+        task: &Task,
+        aggregate_share_req: &AggregateShareReq,
+    ) -> Result<AggregateShareResp, Error>
+    where
+        A: Aggregatable + Send + Sync,
+        E: std::fmt::Debug,
+        Vec<u8>: for<'a> From<&'a A>,
+        for<'a> A: TryFrom<&'a [u8], Error = E>,
+    {
+        // TODO(timg): What should we do if any of the aggregation jobs handling reports relevant to
+        // the aggregate share request aren't finished yet? Helper could return an error like
+        // "EAGAIN" to leader, or could trust leader not to issue an AggregateShareReq until all of
+        // leader's aggregate jobs are done (which should imply that helper's are done).
+
+        let total_aggregate_share = datastore
+            .run_tx(move |tx| {
+                let task = task.clone();
+                let aggregate_share_req = aggregate_share_req.clone();
+                Box::pin(async move {
+                    let batch_aggregations = tx
+                        .get_batch_aggregations::<A, E>(task.id, aggregate_share_req.batch_interval)
+                        .await?;
+
+                    let mut total_report_count = 0;
+                    let mut total_checksum = [0u8; 32];
+                    let mut total_aggregate_share: Option<A> = None;
+
+                    for batch_aggregation in &batch_aggregations {
+                        // TODO(timg): This satisfies the max lifetime requirement in §4.6 but it
+                        // isn't right. We should allow multiple aggregate share requests provided
+                        // the batch interval and aggregation parameter don't change. In fact we
+                        // *must* allow this so that the leader can retry an aggregate share request
+                        // should something go wrong transmitting the results from helper to leader.
+                        // I think we have to amend the protocol text to allow this.
+                        if batch_aggregation.times_collected >= task.max_batch_lifetime {
+                            return Ok(Err(Error::BatchLifetimeExceeded(task.id)));
+                        }
+
+                        // §4.4.4.3: XOR this batch interval's checksum into the overall checksum
+                        total_checksum
+                            .iter_mut()
+                            .zip(batch_aggregation.checksum)
+                            .for_each(|(x, y)| *x ^= y);
+
+                        // §4.4.4.3: Sum all the report counts
+                        total_report_count += batch_aggregation.report_count;
+
+                        match &mut total_aggregate_share {
+                            Some(share) => {
+                                if let Err(err) = share.merge(&batch_aggregation.aggregate_share) {
```

Comment:
VDAFs are defined with an "out_shares_to_agg_share" method, which takes an array of output shares, and produces an aggregate share. The end of that paragraph notes that "Some VDAFs may also support a 'streaming' form, where shares are processed one at a time." A non-streaming VDAF would require a different approach, as you note, but I think the benefits of accumulating aggregation jobs, then batches, then collect requests incrementally are worth it. (For example, a non-streaming VDAF's out_shares_to_agg_share may have a working set size that scales with the number of reports)

Reply:
Right, we can't use `prio::vdaf::Aggregator::aggregate` here because we are not working with output shares, but rather accumulations. `prio::vdaf::Collector::unshard` is what you're meant to use to unshard a sequence of `AggregateShare`s into an `AggregateResult`, so IMO this usage is precisely what `Aggregatable::merge` is for (I'm pretty sure I added `Aggregatable::merge` for exactly this purpose). However it sounds like we could use some better documentation in `libprio-rs` here, and maybe we want some corresponding definitions in the VDAF spec.